### PR TITLE
core: enable fluent bit to do a clean shutdown

### PR DESCRIPTION
**Pull request by epsteina16**
In cases where Fluent Bit has a limited lifetime as a process, it may receive logs and not be able to flush them because an output connection is down. When the engine is shutting down it should try and flush these logs until the grace period is over. Previous behavior made it so task retries were not handled once the engine was shutting down.

Preventing log loss is extremely important for using fluent bit for audit logs.

The changes made include:
- Running task count now reports tasks with scheduled retries
- All retries are rescheduled immediately when the engine first processes a shutdown request
- If grace is set to -1, fluent bit will not finish shutting down until all tasks are flushed
- Retry tasks created during shutdown don't use backoff_full_jitter

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**

Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [N/A] Attached [local packaging test](./packaging/local-build-all.sh) output showing all targets (including any new ones) build.

**Documentation**

- [x] Documentation required for this feature

**Backporting**

- [N/A] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

[test_files.zip](https://github.com/fluent/fluent-bit/files/8288401/test_files.zip)

### CODE REVIEW

File: src/flb_engine.c
Comment (leonardo-albertovich) [inline src/flb_engine.c:127]: This doesn't seem to be necessary right?

File: src/flb_scheduler.c
Comment (leonardo-albertovich) [inline src/flb_scheduler.c:290]: I think part of the coding style is that those variables that are used to hold boolean values using the FLB_TRUE and FLB_FALSE constants are used straight in if blocks without comparing them to the value (ie. `if (myvar)` or `if (!myvar)`) (I like explicit comparisons better but the other way is the standard and having an homogeneous codebase is important)

File: src/flb_scheduler.c
Comment (leonardo-albertovich) [inline src/flb_scheduler.c:702]: You need to release `timer` using `flb_sched_timer_destroy` in this case.

**Review by leonardo-albertovich — CHANGES_REQUESTED** [review]
The code looks good and I agree that it's necessary, the only things I'd change would be renaming `flb_sched_retry_now` and having it take the delay as an argument and then decoupling `flb_engine_reschedule_retries` into the three different levels required to achieve the same and having all those functions work as a chain that takes and forwards a delay value.

That way this code would be ready to be used if the need arises and hopefully it'd make someone elses life easier in the future (and ours as well if it prevents code duplication) but this is definitely not a dealbreaker in my opinion as the change would be simple enough for someone else to do in the future if needed.

The missing cleanup change is the only hard requirement on my side.
Btw, I reviewed the code but did not test this so if you could provide testing instructions that cover all the use cases that would be great (I think at the moment that only means workers = 0 and workers > 0) as I can't remember if there were any particularities when it comes to retry manipulation and worker threads.

**Comment by epsteina16** [comment]
> The code looks good and I agree that it's necessary, the only things I'd change would be renaming `flb_sched_retry_now` and having it take the delay as an argument and then decoupling `flb_engine_reschedule_retries` into the three different levels required to achieve the same and having all those functions work as a chain that takes and forwards a delay value.
> 
> That way this code would be ready to be used if the need arises and hopefully it'd make someone elses life easier in the future (and ours as well if it prevents code duplication) but this is definitely not a dealbreaker in my opinion as the change would be simple enough for someone else to do in the future if needed.
> 
> The missing cleanup change is the only hard requirement on my side. Btw, I reviewed the code but did not test this so if you could provide testing instructions that cover all the use cases that would be great (I think at the moment that only means workers = 0 and workers > 0) as I can't remember if there were any particularities when it comes to retry manipulation and worker threads.

The cleanup change suggestion is good; I'll make that change.
Regarding testing, I attached a zipped test in the pull request. I only tested workers > 0, so I'll add workers=0 to this as well.

**Comment by epsteina16** [comment]
New tests with workers specified are in attached zip. See test-clean-shutdown.sh
[test_files.zip](https://github.com/fluent/fluent-bit/files/8508457/test_files.zip)

**Comment by leonardo-albertovich** [comment]
Excellent! Thank you very much @epsteina16. I'll take a few minutes to test it tomorrow but code wise it looks good @edsiper.

**Comment by epsteina16** [comment]
@leonardo-albertovich Are there any more changes needed on this?

**Review by leonardo-albertovich — DISMISSED** [review]

**Comment by leonardo-albertovich** [comment]
Code wise it looks good to me, do you think you'd be able to turn those two test cases you created into runtime tests? if you check `tests/runtime/out_exit.c` you will find examples. 

In that case what I would do would be setting the `tcp` output plugin just the way you did (one case with workers and one without (which could be implemented in the same function and depend on a parameter of course)) and also set `dummy` as input with the `samples` [option ](https://docs.fluentbit.io/manual/pipeline/inputs/dummy)to generate a finite number of samples.

I think if you do that we would be ready to go. Just to clarify, I think the code is fine as is.

**Comment by epsteina16** [comment]
Sure, that's a good idea. I'll add that to this pull request shortly.

**Comment by epsteina16** [comment]
@leonardo-albertovich I added runtime tests for this, please take a look when you get a chance.

**Comment by leonardo-albertovich** [comment]
Awesome, to be honest, it's been so long (context switch wise) that I completely forgot about this so I'll need a bit of time to review it once again. One question that arose while re-reading your PR message to get into context was this : 

Considering that if `grace` is set to `-1` fluent-bit won't exit until all chunks are flushed and retry tasks are completed is there a "post shutdown retry limit" that would cause fluent-bit to just leave the chunk down (not discard it) after a certain amount of retries in order to be able to shut down? 

I wonder what would happen if the user sets the regular retry limit to `unlimited` and there's an endpoint that's down, maybe it would be a good idea to have an option to limit that, don't you think? Of course it could already be there and I could just be missing the point due to being so out of context, if that's the case then please disregard this.

I'll take a deeper look at this on Monday, thanks a lot for the effort!

**Comment by epsteina16** [comment]
> Considering that if `grace` is set to `-1` fluent-bit won't exit until all chunks are flushed and retry tasks are completed is there a "post shutdown retry limit" that would cause fluent-bit to just leave the chunk down (not discard it) after a certain amount of retries in order to be able to shut down?
> 
> I wonder what would happen if the user sets the regular retry limit to `unlimited` and there's an endpoint that's down, maybe it would be a good idea to have an option to limit that, don't you think? Of course it could already be there and I could just be missing the point due to being so out of context, if that's the case then please disregard this.

In my opinion, that is the point of the retry_limit setting for output plugins. If retry_limit is set to false, and grace is set to -1, when fluent bit is killed the user will get exactly what they want - unlimited retries until success and then shutdown.

**Comment by leonardo-albertovich** [comment]
I'm not sure about that, I mean, if your endpoint is down and you want to shut fluent bit down for maintenance or whatever then there has to be some sort of limit, otherwise you'll end up force killing it which is not good at all. What I'm trying to describe doesn't involve dropping the chunks or not giving them a chance to flush but giving the user a way to limit that.

I understand what you mean about that being the expected behavior but I think it's a slippery slope and would like to consider all options while at it.

**Comment by epsteina16** [comment]
You are certainly right that it is a slippery slope. We could add a large, hard to hit hard-timeout (10 minutes?). 

Otherwise providing another setting for a retry or time-based limit when set to grace -1 seems unnecessary to me, since grace could just be set to a large positive number and retry_limit could be set to a large number as well.

**Comment by leonardo-albertovich** [comment]
Maybe we could fix that adding periodical warnings every N minutes after a shutdown is blocked by retries if both things are set to unlimited, would you be OK with that? (I think maybe N = 1, 2 or 5 would work)

**Comment by epsteina16** [comment]
Yes, that seems reasonable to me. I will add that to this PR.

**Comment by leonardo-albertovich** [comment]
Awesome, once you do we'll be ready to merge, thanks a lot!

**Comment by epsteina16** [comment]
I went back and looked at the changes and there already is a warning message that is triggered every 1 second as part of the engine shutdown loop. 
https://github.com/fluent/fluent-bit/blob/3937e43f87236bc494ec625e324ec64f43600da3/src/flb_engine.c#L843

**Comment by leonardo-albertovich** [comment]
Coolio, I've scheduled time to review everything again today. Once I do that I'll approve it and it'll be merged soon. 

Thanks for the cooperation! ( and for the patience =P )

**Comment by epsteina16** [comment]
@leonardo-albertovich just checking if there is anything else you need from me for this?

**Review by leonardo-albertovich — APPROVED** [review]

**Comment by leonardo-albertovich** [comment]
It seems like all of them are happy except two that are not up to the PR to fix so I'm cool with merging this as is @edsiper @niedbalski @patrick-stephens. Please let me know if you need anything.

Thanks a lot @epsteina16!

**Review by leonardo-albertovich — APPROVED** [review]

**Comment by patrick-stephens** [comment]
Agreed clean shutdown with flush will be a great improvement for automated testing and other batch processing use cases.

**Review by niedbalski — APPROVED** [review]

**Pull request merged**
